File: TwoPointers/CPP/long_pressed_name.cpp
#include <string>
using namespace std;

class Solution {
public:
    bool isLongPressedName(string name, string typed) {
        int i = 0,
            j = 0,
            cntn = 0,
            cntt = 0,
            m = name.size(),
            n = typed.size();
        char cur = name[0];
        if (name[0] != typed[0]) return false;
        while (i < m) {
            char cur = name[i];
            if (cur != typed[j]) return false;
            while (j < n && typed[j] == cur) {
                cntt++;
                j++;
            }
            while (i < m && name[i] == cur) {
                cntn++;
                i++;
            }
            if (cntt < cntn || (j < n && i >= m) || j >= n && i < m) return false;
            cntn = 0;
            cntt = 0;
        }
        return true;
    }
};

File: TwoPointers/long_pressed_name.py
class Solution:
    def isLongPressedName(self, name: str, typed: str) -> bool:
        i, j = 0, 0
        cntn, cntt = 0, 0
        cur = name[0]
        m, n = len(name), len(typed)
        if name[0] != typed[0]: return False
        while i < m:
            cur = name[i]
            if cur != typed[j]: return False
            while j < n and typed[j] == cur:
                cntt += 1
                j += 1
            while i < m and name[i] == cur:
                cntn += 1
                i += 1
            if cntt < cntn or (j < n and i >= m) or (j >= n and i < m): return False
            cntn, cntt = 0, 0
        return True
        